modules/javafx.graphics/src/main/native-glass/gtk/glass_dnd.cpp

```diff
@@ -760,132 +734,346 @@ static gboolean dnd_source_set_raw(GtkWidget *widget, GtkSelectionData *sel_data
+    }
+
+    g_list_free(targets);
+
+    gboolean is_raw_image = FALSE;
+    gint w = 0, h = 0;
+    GdkPixbuf* pixbuf = DragView::get_drag_image(data, &is_raw_image, &w, &h);
+    gint offset_x = -2;
+    gint offset_y = -2;
 
-    // this widget is used only to pass events and will
-    // be destroyed on drag end
-    drag_widget = gtk_window_new(GTK_WINDOW_POPUP);
-    gtk_window_resize(GTK_WINDOW(drag_widget), 1, 1);
-    gtk_window_move(GTK_WINDOW(drag_widget), -200, -200);
-    gtk_widget_show(drag_widget);
+    if (GDK_IS_PIXBUF(pixbuf)) {
+        offset_x = w / 2;
+        offset_y = h / 2;
 
-    g_object_set_data_full(G_OBJECT(drag_widget), SOURCE_DND_DATA, data, clear_global_ref);
+        DragView::get_drag_image_offset(data, &offset_x, &offset_y);
+    } else {
+        GError *error = NULL;
+        GtkIconTheme *icon_theme = gtk_icon_theme_get_default();
+        gtk_icon_size_lookup(GTK_ICON_SIZE_DND, &w, &h);
+        pixbuf = gtk_icon_theme_load_icon(icon_theme,
+                                         "text-x-generic", w,
+                                         GTK_ICON_LOOKUP_USE_BUILTIN,
+                                         &error);
+    }
 
-    g_signal_connect(drag_widget, "drag-begin",
-        G_CALLBACK(dnd_drag_begin_callback), NULL);
+    if (GDK_IS_PIXBUF(pixbuf)) {
+        ctx->drag_view = new DragView(pixbuf, is_raw_image, offset_x, offset_y);
+    }
+}
 
-    g_signal_connect(drag_widget, "drag-failed",
-        G_CALLBACK(dnd_drag_failed_callback), NULL);
+//true = continue
+bool process_dnd_source(GdkEvent *event) {
+    GdkWindow* window = event->any.window;
 
-    g_signal_connect(drag_widget, "drag-data-get",
-        G_CALLBACK(dnd_data_get_callback), NULL);
+    if (!in_drag || !window) {
+        return true;
+    }
 
-    g_signal_connect(drag_widget, "drag-end",
-        G_CALLBACK(dnd_end_callback), NULL);
+    DragView *drag_view = (DragView *) g_object_get_data(G_OBJECT(window), SOURCE_DND_DRAG_VIEW);
+    if (drag_view != NULL && event->type == GDK_EXPOSE && window == drag_view->get_window()) {
+        drag_view->expose();
+        return true;
+    }
 
-    GtkTargetList *tlist = data_to_gtk_target_list(env, data);
+    DragSourceContext *ctx = (DragSourceContext*)
+        g_object_get_data(G_OBJECT(window), SOURCE_DND_CONTEXT);
 
-    GdkDragContext *context;
+    if (!ctx) {
+        return true;
+    }
 
-    gint x, y;
-    glass_gdk_master_pointer_get_position(&x, &y);
+    if (event->type == GDK_DELETE && window == ctx->dnd_window) {
+        gdk_drag_abort(ctx->dnd_ctx, GDK_CURRENT_TIME);
+        in_drag = false;
+        return true;
+    }
 
-    is_dnd_owner = TRUE;
+    // the window has grab
+    if (window != ctx->dnd_window) {
+        return true;
+    }
 
-    context = gtk_drag_begin(drag_widget, tlist, actions, 1, NULL);
+    switch(event->type) {
+        case GDK_GRAB_BROKEN:
+            process_dnd_source_grab_broken(ctx, event);
+            break;
```

Comment:
This was added to consider the case when a mouse grab is broken, I just could not think when it happens - but it's present on gdk managed drag and drop code.